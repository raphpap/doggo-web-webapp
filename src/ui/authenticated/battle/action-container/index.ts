export * from './action-container';
export {default} from './action-container';
