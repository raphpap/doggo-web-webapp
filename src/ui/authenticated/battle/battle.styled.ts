// Vendor
import styled from 'react-emotion';

export const ActionContainer = styled.button`
  width: 160px;
  height: 160px;
  border: 2px solid #fff;
  border-radius: 50%;
  margin: 60px 0;
  background-color: transparent;
  outline: none;
  font-size: 16px;
  color: #fff;
  cursor: pointer;
`;

export const CardContainer = styled.ul`
  min-width: 320px;
  max-width: 800px;
  padding: 0;
`;

export const Message = styled.p`
  color: rgba(255, 255, 255, 0.8);
`;

export const CardSelectionButton = styled.button`
  width: 400px;
  height: 80px;
  border: 1px solid #fff;
  margin-top: 16px;
  background-color: transparent;
  font-size: 24px;
  color: #fff;
  cursor: pointer;
`;

export const ModalCss = {
  closeIcon: {
    fill: 'rgba(255, 255, 255, 0.8)'
  },
  modal: {
    backgroundColor: '#171820',
    paddingTop: '40px'
  }
};
