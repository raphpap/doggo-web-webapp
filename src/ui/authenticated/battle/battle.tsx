// Vendor
import React from 'react';
import {compose} from 'recompose';

// Context
import {
  Card,
  withApplicationContext,
  WithApplicationContextProps
} from 'doggo/context';

// Shared Components
import BigCardModal from 'doggo/ui/@components/big-card-modal';
import Message from 'doggo/ui/@components/message';

// Components
import CardSelectionModal from './card-selection-modal';
import Container from './container';
import NextBattleAction from './next-battle-action';
import OpponentCard from './opponent-card';
import OwnCard from './own-card';

// Utilities
import findCard from 'doggo/utilities/find-card';

// Types
interface Props {}
type EnhancedProps = Props & WithApplicationContextProps;

interface State {
  showCardSelectionModal: boolean;
  showModalCard: Card | null;
}

const enhance = compose<EnhancedProps, Props>(withApplicationContext);

export class Battle extends React.Component<EnhancedProps, State> {
  public readonly state: State = {
    showCardSelectionModal: false,
    showModalCard: null
  };

  public render() {
    const {showCardSelectionModal, showModalCard} = this.state;
    const {cards, battle} = this.props.context.state;
    const {opponent, cardId: selectedCardId} = battle;

    if (!cards || !opponent) return <Message>Loading...</Message>;

    const selectedCard = findCard(selectedCardId || '', cards) || null;

    return (
      <div>
        <Container>
          <OpponentCard
            opponentCard={opponent}
            onCardClick={this.openCardModal}
          />

          <NextBattleAction
            ownCard={selectedCard}
            battleStatus={battle.status}
            onBattleClicked={this.handleBattleClicked}
            onNextAfterLost={this.handleNextAfterLost}
            onNextAfterWin={this.handleNextAfterWin}
          />

          <OwnCard
            ownCard={selectedCard}
            onCardClick={this.openCardModal}
            onCardSelectionClick={this.openCardSelectionModal}
          />
        </Container>

        <CardSelectionModal
          isOpen={showCardSelectionModal}
          onClose={this.closeCardSelectionModal}
          cards={cards}
          onCardSelect={this.handleCardSelected}
        />

        <BigCardModal card={showModalCard} onClose={this.closeCardModal} />
      </div>
    );
  }

  private handleNextAfterLost = () => {
    const {unselectBattleCard} = this.props.context.actions;
    unselectBattleCard();
  };

  private handleNextAfterWin = () => {
    const {opponent} = this.props.context.state.battle;
    const {getNextOpponent} = this.props.context.actions;

    if (opponent) {
      getNextOpponent({card: opponent});
    }
  };

  private handleBattleClicked = (selectedCard: Card) => {
    const {battle: battleAction} = this.props.context.actions;
    const {battle} = this.props.context.state;
    const {opponent} = battle;

    if (selectedCard && opponent) {
      battleAction({ownCard: selectedCard, opponentCard: opponent});
    }
  };

  private handleCardSelected = (card: Card) => {
    const {selectBattleCard} = this.props.context.actions;
    selectBattleCard(card);
    this.setState({showCardSelectionModal: false});
  };

  private openCardSelectionModal = () => {
    this.setState({showCardSelectionModal: true});
  };

  private closeCardSelectionModal = () => {
    this.setState({showCardSelectionModal: false});
  };

  private openCardModal = (card: Card) => {
    this.setState({showModalCard: card});
  };

  private closeCardModal = () => {
    this.setState({showModalCard: null});
  };
}

export default enhance(Battle);
