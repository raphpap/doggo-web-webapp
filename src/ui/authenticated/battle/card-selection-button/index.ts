export * from './card-selection-button';
export {default} from './card-selection-button';
