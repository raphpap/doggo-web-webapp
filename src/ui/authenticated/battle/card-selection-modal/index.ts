export * from './card-selection-modal';
export {default} from './card-selection-modal';
