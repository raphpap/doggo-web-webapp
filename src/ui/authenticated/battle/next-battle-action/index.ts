export * from './next-battle-action';
export {default} from './next-battle-action';
