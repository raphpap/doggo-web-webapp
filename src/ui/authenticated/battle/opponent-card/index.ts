export * from './opponent-card';
export {default} from './opponent-card';
