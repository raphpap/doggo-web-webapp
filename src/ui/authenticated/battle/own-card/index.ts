export * from './own-card';
export {default} from './own-card';
