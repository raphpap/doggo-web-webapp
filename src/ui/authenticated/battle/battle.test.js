// Vendor
import React from 'react';
import {shallow} from 'enzyme';
import toJson from 'enzyme-to-json';

// Components
import CardSelectionModal from './card-selection-modal';
import {Battle} from './battle';
import NextBattleAction from './next-battle-action';
import OpponentCard from './opponent-card';
import OwnCard from './own-card';

// Shared Components
import BigCardModal from 'doggo/ui/@components/big-card-modal';

// Mocks
import {mockCard, mockLoggedIn} from 'doggo/test-mocks';

describe('<Battle />', () => {
  let wrapper;
  let context;
  const card11 = mockCard(11);
  const card12 = mockCard(12);

  beforeEach(() => {
    context = mockLoggedIn();
    context.state.cards = [card11, card12];
    wrapper = shallow(<Battle context={context} />);
  });

  it('renders correctly', () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('renders the opponent card correctly', () => {
    const opponentCard = wrapper.find(OpponentCard);
    expect(opponentCard).toBeTruthy();
    expect(opponentCard.props().opponentCard).toBe(context.state.battle.opponent);
  });

  it('renders the next battle action correctly when no selected card', () => {
    const nextBattleAction = wrapper.find(NextBattleAction);
    expect(nextBattleAction).toBeTruthy();
    expect(nextBattleAction.props().ownCard).toBe(null);
  });

  it('renders the next battle action correctly when there is a selected card', () => {
    context.state.battle.cardId = '12';
    wrapper.setProps({context})

    const nextBattleAction = wrapper.find(NextBattleAction);
    expect(nextBattleAction).toBeTruthy();
    expect(nextBattleAction.props().ownCard).toBe(card12);
  });

  it('renders the own card correctly when no selected card', () => {
    const ownCard = wrapper.find(OwnCard);
    expect(ownCard).toBeTruthy();
    expect(ownCard.props().ownCard).toBe(null);
  });

  it('renders the own card correctly when there is a selected card', () => {
    context.state.battle.cardId = '12';
    wrapper.setProps({context})

    const ownCard = wrapper.find(OwnCard);
    expect(ownCard).toBeTruthy();
    expect(ownCard.props().ownCard).toBe(card12);
  });

  it('renders the CardSelectionModal correctly when opened', () => {
    wrapper.setState({showCardSelectionModal: true})

    const cardSelectionModal = wrapper.find(CardSelectionModal);
    expect(cardSelectionModal).toBeTruthy();
    expect(cardSelectionModal.props().isOpen).toBe(true);
    expect(cardSelectionModal.props().cards).toBe(context.state.cards);
  });

  it('renders the CardSelectionModal correctly when opened', () => {
    wrapper.setState({showCardSelectionModal: false})

    const cardSelectionModal = wrapper.find(CardSelectionModal);
    expect(cardSelectionModal).toBeTruthy();
    expect(cardSelectionModal.props().isOpen).toBe(false);
    expect(cardSelectionModal.props().cards).toBe(context.state.cards);
  });

  it('renders the BigCardModal correctly when no card to show', () => {
    const bigCardModal = wrapper.find(BigCardModal);
    expect(bigCardModal).toBeTruthy();
    expect(bigCardModal.props().card).toBe(null);
  });

  it('renders the BigCardModal correctly when there is a card to show', () => {
    wrapper.setState({showModalCard: card12});
    const bigCardModal = wrapper.find(BigCardModal);
    expect(bigCardModal).toBeTruthy();
    expect(bigCardModal.props().card).toBe(card12);
  });
});
