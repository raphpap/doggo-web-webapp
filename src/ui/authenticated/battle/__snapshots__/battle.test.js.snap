// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Battle /> renders correctly 1`] = `
<Wrapper>
  <Container>
    <OpponentCard
      onCardClick={[Function]}
      opponentCard={
        Object {
          "attack": 99,
          "defense": 99,
          "hpLeft": 99,
          "hpTotal": 99,
          "id": "99",
          "image": "99",
          "name": "99",
        }
      }
    />
    <NextBattleAction
      battleStatus="no-card-selected"
      onBattleClicked={[Function]}
      onNextAfterLost={[Function]}
      onNextAfterWin={[Function]}
      ownCard={null}
    />
    <OwnCard
      onCardClick={[Function]}
      onCardSelectionClick={[Function]}
      ownCard={null}
    />
  </Container>
  <CardSelectionModal
    cards={
      Array [
        Object {
          "attack": 11,
          "defense": 11,
          "hpLeft": 11,
          "hpTotal": 11,
          "id": "11",
          "image": "11",
          "name": "11",
        },
        Object {
          "attack": 12,
          "defense": 12,
          "hpLeft": 12,
          "hpTotal": 12,
          "id": "12",
          "image": "12",
          "name": "12",
        },
      ]
    }
    isOpen={false}
    onCardSelect={[Function]}
    onClose={[Function]}
  />
  <BigCardModal
    card={null}
    onClose={[Function]}
  />
</Wrapper>
`;
